### PR TITLE
Density Function Fixes and Improvements

### PR DESCRIPTION
**Pull request by jacobsjo**
a few fixes and improvements for the density functions. I'll explain each change separately in comments.

### CODE REVIEW

File: src/worldgen/DensityFunction.ts [hunk 1]
```diff
@@ -64,9 +64,9 @@ export namespace DensityFunction {
 		const root = Json.readObject(obj) ?? {}
 		const type = Json.readString(root.type)?.replace(/^minecraft:/, '')
 		switch (type) {
-			case 'blend_alpha': return Constant.ONE
-			case 'blend_offset': return Constant.ZERO
-			case 'beardifier': return Constant.ZERO
+			case 'blend_alpha': return new ConstantMinMax(1, 0, 1)
+			case 'blend_offset': return new ConstantMinMax(0, -Infinity, Infinity)
+			case 'beardifier': return new ConstantMinMax(0, -Infinity, Infinity)
```
Comment (jacobsjo) [inline src/worldgen/DensityFunction.ts:69]: Even though these are constant for us, it makes sense to have the min and max values represent the ones in vanilla.
(on a second thought, maybe blend_alpha and blend_offset are only replaced if there is some blending to do?)

File: src/worldgen/DensityFunction.ts [hunk 2]
```diff
@@ -307,21 +325,21 @@ export namespace DensityFunction {
 		public compute({ x: blockX, y: blockY, z: blockZ }: DensityFunction.Context) {
 			const w = this.cellWidth
 			const h = this.cellHeight
-			const x = (blockX % w) / w
-			const y = (blockY % h) / h
-			const z = (blockZ % w) / w
+			const x = ((blockX % w + w) % w) / w
+			const y = ((blockY % h + h) % h) / h
+			const z = ((blockZ % w + w) % w) / w
```
Comment (jacobsjo) [inline src/worldgen/DensityFunction.ts:330]: stupid JavaScript modulo is what caused the interpolated to misbehave.

File: src/worldgen/DensityFunction.ts [hunk 3]
```diff
@@ -307,21 +325,21 @@ export namespace DensityFunction {
...
-			return lerp3(x, y, z, noise000, noise100, noise010, noise110, noise001, noise101, noise011, noise111)
+			const noise000 = () => this.computeCorner(firstX, firstY, firstZ)
+			const noise001 = () => this.computeCorner(firstX, firstY, firstZ + w)
+			const noise010 = () => this.computeCorner(firstX, firstY + h, firstZ)
+			const noise011 = () => this.computeCorner(firstX, firstY + h, firstZ + w)
+			const noise100 = () => this.computeCorner(firstX + w, firstY, firstZ)
+			const noise101 = () => this.computeCorner(firstX + w, firstY, firstZ + w)
+			const noise110 = () => this.computeCorner(firstX + w, firstY + h, firstZ)
+			const noise111 = () => this.computeCorner(firstX + w, firstY + h, firstZ + w)
+			return lazyLerp3(x, y, z, noise000, noise100, noise010, noise110, noise001, noise101, noise011, noise111)
```
Comment (jacobsjo) [inline src/worldgen/DensityFunction.ts:342]: especially when only calculating individual slices, we don't need to calculate 8 values for each position, 4 are enough, so this lazyLerp doesn't calculate the corners it doesn't need.

File: src/worldgen/DensityFunction.ts [hunk 4]
```diff
@@ -467,7 +485,7 @@ export namespace DensityFunction {
 			return visitor(new RangeChoice(this.input.mapAll(visitor), this.minInclusive, this.maxExclusive, this.whenInRange.mapAll(visitor), this.whenOutOfRange.mapAll(visitor)))
 		}
 		public minValue() {
-			return Math.min(this.whenInRange.minValue(), this.whenOutOfRange.maxValue())
+			return Math.min(this.whenInRange.minValue(), this.whenOutOfRange.minValue())
```
Comment (jacobsjo) [inline src/worldgen/DensityFunction.ts:488]: a simple typo

File: src/worldgen/DensityFunction.ts [hunk 5]
```diff
@@ -528,7 +546,7 @@ export namespace DensityFunction {
 			super(noiseData, offsetNoise)
 		}
 		public withNewNoise(newNoise: NormalNoise) {
-			return new ShiftB(this.noiseData, newNoise)
+			return new Shift(this.noiseData, newNoise)
```
Comment (jacobsjo) [inline src/worldgen/DensityFunction.ts:549]: another typo

File: src/worldgen/DensityFunction.ts [hunk 6]
```diff
@@ -697,12 +715,12 @@ export namespace DensityFunction {
...
-							: Math.min(min1 * max2, min2 * max1)
-					max = min1 > 0 && min2 > 0 ? max1 * max2
-						: max1 < 0 && max2 < 0 ? min1 * min2
-							: Math.max(min1 * min2, max1 * max2)
+					min = min1 > 0 && min2 > 0 ? (min1 * min2) || 0
+						: max1 < 0 && max2 < 0 ? (max1 * max2) || 0
+							: Math.min((min1 * max2) || 0, (min2 * max1) || 0)
+					max = min1 > 0 && min2 > 0 ? (max1 * max2) || 0
+						: max1 < 0 && max2 < 0 ? (min1 * min2) || 0
+							: Math.max((min1 * min2) || 0, (max1 * max2) || 0)
```
Comment (jacobsjo) [inline src/worldgen/DensityFunction.ts:723]: `Infinity * 0 = NaN`. usually we want the result of that multiplication to be 0 though. I think this "bug" exists in vanilla Minecraft as well, and it doesn't hurt anything except performance, but having NaN as the min and max values everywhere is ugly...

File: src/worldgen/NoiseRouter.ts [hunk 7]
```diff
@@ -109,6 +109,9 @@ export namespace NoiseRouter {
 			if (fn instanceof DensityFunction.OldBlendedNoise) {
 				return new DensityFunction.OldBlendedNoise(new BlendedNoise(random.fromHashOf(Identifier.create('terrain').toString()), settings.sampling, NoiseSettings.cellWidth(settings), NoiseSettings.cellHeight(settings)))
 			}
+			if (fn instanceof DensityFunction.Mapped){
+				return fn.withMinMax()
+			}
```
Comment (jacobsjo) [inline src/worldgen/NoiseRouter.ts:114]: min and max of mapped types weren't calculated properly.

File: src/worldgen/DensityFunction.ts [hunk 8]
```diff
@@ -590,7 +608,7 @@ export namespace DensityFunction {
 		}
 		private readonly transformer: (density: number) => number
 		constructor(
-			private readonly type: typeof MappedType[number],
+			public readonly type: typeof MappedType[number],
```
Comment (jacobsjo) [inline src/worldgen/DensityFunction.ts:611]: I needed access to the type, and since the type of AP2 is also public...

**Comment by misode** [comment]
Changes all look good! Thanks a lot for spotting all these issues!

**Pull request merged**